near-contract-standards/src/fungible_token/metadata.rs

```diff
@@ -1,20 +1,29 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
 use near_sdk::json_types::Base64VecU8;
-use near_sdk::serde::Serialize;
+use near_sdk::serde::{Deserialize, Serialize};
 
-pub const FT_METADATA_SPEC: &str = "ftm-1.0.0";
+pub const FT_METADATA_SPEC: &str = "ft-1.0.0";
 
-#[derive(Serialize)]
+#[derive(BorshDeserialize, BorshSerialize, Clone, Deserialize, Serialize)]
 #[serde(crate = "near_sdk::serde")]
 pub struct FungibleTokenMetadata {
     pub spec: String,
     pub name: String,
     pub symbol: String,
     pub icon: Option<String>,
-    pub reference: String,
-    pub reference_hash: Base64VecU8,
+    pub reference: Option<String>,
+    pub reference_hash: Option<Base64VecU8>,
     pub decimals: u8,
 }
 
 pub trait FungibleTokenMetadataProvider {
     fn ft_metadata(&self) -> FungibleTokenMetadata;
 }
+
+impl FungibleTokenMetadata {
+    pub fn assert_valid(&self) {
```

Comment:
I think we also need to make sure that someone doesn't just throw in `reference` and expect it to be there. I was under the impression that if you want to use the Metadata standard, you have to include the required fields (`spec`, `name`, `symbol`, `decimals`) and we wouldn't allow them to disregard those.

Reply:
These are required fields on the structure. What we can enforce is if `reference` is given, then we require `reference_hash`

Reply:
I also forgot to enforce SPEC

Reply:
Ah I see. Yes, it'll panic and tell you the next field that you forgot to add.

`message: 'Smart contract panicked: panicked at &#39;Failed to deserialize input from JSON.: Error(&quot;missing field &#x60;symbol&#x60;&quot;, line: 1, column: 117)&#39;, ft&#x2F;src&#x2F;lib.rs:35:1',`
even though I also don't have decimals.

Doesn't tell you "you forgot these fields" which isn't as great, but I'm not going to be a stick in the mud.